common/src/main/java/com/microsoft/identity/common/internal/controllers/NativeAuthMsalController.kt

```diff
@@ -334,6 +350,167 @@ class NativeAuthMsalController : BaseNativeAuthController() {
+                    )
+                }
+                is SignUpStartApiResult.InvalidEmail -> {
+                    SignUpCommandResult.InvalidEmail(
+                        error = signUpStartApiResult.error,
+                        errorDescription = signUpStartApiResult.errorDescription
+                    )
+                }
+                is SignUpStartApiResult.AuthNotSupported -> {
+                    SignUpCommandResult.AuthNotSupported(
+                        error = signUpStartApiResult.error,
+                        errorDescription = signUpStartApiResult.errorDescription
+                    )
+                }
+                is SignUpStartApiResult.Redirect -> {
+                    INativeAuthCommandResult.Redirect()
+                }
+                is SignUpStartApiResult.UnsupportedChallengeType, is SignUpStartApiResult.UnknownError -> {
+                    signUpStartApiResult as ApiErrorResult
+                    Logger.warn(
+                        TAG,
+                        "Unexpected result: $signUpStartApiResult"
+                    )
+                    INativeAuthCommandResult.UnknownError(
+                        error = signUpStartApiResult.error,
+                        errorDescription = signUpStartApiResult.errorDescription
+                    )
+                }
+            }
+        } catch (e: Exception) {
+            Logger.error(TAG, "Exception thrown in signUpStart", e)
+            throw e
+        }
+    }
+
+    /**
+     * Makes a call to the signup/continue endpoint using the provided code, mapping responses returned from the server into a command result.
+     */
+    fun signUpSubmitCode(parameters: SignUpSubmitCodeCommandParameters): SignUpSubmitCodeCommandResult {
+        LogSession.logMethodCall(TAG, "${TAG}.signUpSubmitCode")
+
+        try {
+            val oAuth2Strategy = createOAuth2Strategy(parameters)
+
+            val signUpContinueApiResult = performSignUpSubmitCode(
+                oAuth2Strategy = oAuth2Strategy,
+                parameters = parameters
+            )
+            return signUpContinueApiResult.toSignUpSubmitCodeCommandResult(oAuth2Strategy)
+        } catch (e: Exception) {
+            Logger.error(TAG, "Exception thrown in signUpSubmitCode", e)
+            throw e
+        }
+    }
+
+    /**
+     * Makes a call to the signup/challenge endpoint to trigger a code to be re-sent.
+     */
+    fun signUpResendCode(parameters: SignUpResendCodeCommandParameters): SignUpResendCodeCommandResult {
+        LogSession.logMethodCall(TAG, "${TAG}.signUpResendCode")
+
+        try {
+            val oAuth2Strategy = createOAuth2Strategy(parameters)
+
+            return performSignUpChallengeCall(
+                oAuth2Strategy = oAuth2Strategy,
+                signupToken = parameters.signupToken
+            ).toSignUpStartCommandResult() as SignUpResendCodeCommandResult
```

Comment:
Why do we need to invoke these `to...` methods on the results and transform them? Why can't `performSignUpChallengeCall` just return the desired output? 

Same comment here, above, below and everywhere such pattern is followed

Reply:
This was done to separate the calls into two distinct layers. The conversion of  apiResult to  commandResult fits the structure of when case, it's clearer to split it into two methods without having to go deep inside the network layer.